backend/layers/processing/process_validate_atac.py

```diff
@@ -0,0 +1,219 @@
+        Hash the file
+        :param file_name: the file to hash
+        :return: the hash
+        """
+        with open(file_name, "rb") as f:
+            return hashlib.sha256(f.read()).hexdigest()
+
+    def process(
+        self,
+        collection_version_id: CollectionVersionId,
+        dataset_version_id: DatasetVersionId,
+        manifest: IngestionManifest,
+        datasets_bucket: str,
+    ):
+        """
+
+        :param collection_version_id:
+        :param dataset_version_id:
+        :param manifest:
+        :param datasets_bucket:
+        :return:
+        """
+        # Download the original dataset files from URI
+        local_anndata_filename = self.download_from_source_uri(
+            source_uri=str(manifest.anndata),
+            local_path=CorporaConstants.ORIGINAL_H5AD_ARTIFACT_FILENAME,
+        )
+
+        if self.skip_atac_validation(local_anndata_filename, manifest, dataset_version_id):
+            return
+
+        # Download the original fragment file from URI
+        local_fragment_filename = self.download_from_source_uri(
+            source_uri=str(manifest.atac_fragment), local_path=CorporaConstants.ORIGINAL_ATAC_FRAGMENT_FILENAME
+        )
+
+        # Hash the original fragment file. This is used to check if the new fragment file is the same as the old
+        # fragment file to avoid uploading the same file multiple times
+        if self.business_logic.is_already_ingested(manifest.atac_fragment):
+            original_fragment_hash = self.hash_file(local_fragment_filename)
+        else:
+            # new fragment file is from a private source, so we must upload it
+            original_fragment_hash = None
+
+        # Validate the fragment with anndata file
+        try:
+            errors = self.schema_validator.validate_atac(local_fragment_filename, local_anndata_filename)
+        except Exception as e:
+            # for unexpected errors, log the exception and raise a ValidationAtacFailed exception
+            self.logger.exception("validation failed")
+            self.update_processing_status(
+                dataset_version_id, DatasetStatusKey.ATAC_FRAGMENT, DatasetConversionStatus.FAILED
+            )
+            raise ValidationAtacFailed(errors=[str(e)]) from None
+
+        if errors:
+            # if the validation fails, update the processing status and raise a ValidationAtacFailed exception
+            self.update_processing_status(
+                dataset_version_id, DatasetStatusKey.ATAC_FRAGMENT, DatasetConversionStatus.FAILED
+            )
+            raise ValidationAtacFailed(errors=errors)
+
+        # check to see if the new fragments is the same as the old fragment
+        # if it is the same, skip the upload and use link the old fragment to the new dataset
+        # This case is for when the dataset is reprocessed, or updated
+        if original_fragment_hash is not None and original_fragment_hash == self.hash_file(local_fragment_filename):
```

Comment:
I don't understand the logic here. On line 164 above:

```python
original_fragment_hash = self.hash_file(local_fragment_filename)
```

But it's not immediately obvious to me that that path would be overwritten by now.

I am also not able to figure out what `self.schema_validator.validate_atac` does at the moment (see other comment)

Reply:
Depending on the definition of `validate_atac`, do we actually expect it to return exactly the same result every time? I wouldn't expect deterministic hash values from the processing because you were appending chromosomes of data to the fragment file based on when they were done, not with a global order.

Reply:
True, it may not be consistent. Have any ideas how to get a consistent hash?

Reply:
if could calculate the hash in the cli, we could hash the separate chromosomes and combine those hashes in order to achieve a consistent hash. I don't think it's possible for us to do in the 1 day we have remaining for this work.

Reply:
I would make the append step serial so you can control the order

Reply:
I'm also not actually sure if gzip is deterministic by default. A quick google says you might need to set some flags. Either that or hash the decompressed data.

Reply:
We can create a ticket to do more validation in this area.